cpp/earley_parser.cc

```diff
@@ -31,109 +31,151 @@ void EarleyParser::PopLastStates(int32_t cnt) {
+    for (const auto& [ref_id, parent_state] : tmp_rule_id_to_completable_states_) {
+      if (ref_id != state.rule_id) {
+        continue;
+      }
+      if (parent_state.rule_id == -1 ||
+          !grammar_->per_rule_fsms[parent_state.rule_id].has_value()) {
+        const auto& parent_expr = grammar_->GetGrammarExpr(parent_state.sequence_id);
+        const auto& element_expr = grammar_->GetGrammarExpr(parent_expr[parent_state.element_id]);
+        // The new rule is not referenced by a fsm.
+        XGRAMMAR_DCHECK(
+            element_expr.type == GrammarExprType::kRuleRef ||
+            element_expr.type == GrammarExprType::kRepeat
         );
+        if (element_expr.type == GrammarExprType::kRuleRef) {
+          Enqueue(ParserState{
+              parent_state.rule_id,
+              parent_state.sequence_id,
+              parent_state.element_id + 1,
+              parent_state.rule_start_pos,
+              0
+          });
+          continue;
+        }
+        XGRAMMAR_DCHECK(element_expr.type == GrammarExprType::kRepeat);
+        if (std::binary_search(
+                grammar_->allow_empty_rule_ids.begin(),
+                grammar_->allow_empty_rule_ids.end(),
+                element_expr[0]
+            )) {
+          // It means that the subrule of the repeat is empty, and we have already detected it.
+          // We shouldn't add it into the queue.
+          continue;
+        }
+        // The parent state is a repeat, we need to increase the repeat count.
+        auto new_state = parent_state;
+        const int32_t& min_repeat_count = element_expr[1];
+        const int32_t& max_repeat_count = element_expr[2];
+        new_state.repeat_count++;
+        // The repeat rule can be completed, and we advance the state. Don't forget to
+        // reset the repeat count.
+        if (new_state.repeat_count >= min_repeat_count) {
+          Enqueue(ParserState{
+              parent_state.rule_id,
+              parent_state.sequence_id,
+              parent_state.element_id + 1,
+              parent_state.rule_start_pos,
+              0
+          });
+        }
+        // If the repeat count is less than the max repeat count, we can continue to
+        // visit the repeat state for another round.
+        if (new_state.repeat_count < max_repeat_count) {
+          Enqueue(new_state);
+        }
+        continue;
       }
+      // If the rule is referenced by a fsm, we need to advance the fsm.
+      XGRAMMAR_DCHECK(grammar_->per_rule_fsms[parent_state.rule_id].has_value());
+      Enqueue(parent_state);
     }
   }
 }
 
 std::pair</* scanable */ bool, /* completable */ bool> EarleyParser::Predict(
-    const ParserState& state, const GrammarExpr& grammar_expr
+    const ParserState& state
 ) {
-  // Check if it's the tag dispatch.
+  // Check if the rule has a corresponding FSM.
   if (state.rule_id != -1 && grammar_->per_rule_fsms[state.rule_id].has_value()) {
```

Comment:
Add test for NFA
rule = rule1 | rule2 | rule3 
rule 1 = "abc" | ""
rule 2 = "abd" | ""
rule 3 = [a-n] [b-c] x | ""

Assert accepted